File: src/utils/importer.ts
/**
 * TypeScript importer for the Android/web backend (the desktop build parses in
 * Rust). Supports Bitwarden JSON, KeePass 2.x XML, and the CSV family
 * (Chrome/Edge, Firefox, LastPass, 1Password, generic) via smart header mapping.
 */
import type { NewEntry } from "../types";

export function parseImport(format: string, content: string): NewEntry[] {
  switch (format.toLowerCase()) {
    case "bitwarden":
      return parseBitwarden(content);
    case "keepass":
    case "keepass-xml":
      return parseKeePassXml(content);
    default:
      return parseCsv(content);
  }
}

function parseBitwarden(text: string): NewEntry[] {
  let json: unknown;
  try {
    json = JSON.parse(text);
  } catch {
    throw new Error("Ungültige Bitwarden-JSON");
  }
  const items = (json as { items?: unknown[] }).items;
  if (!Array.isArray(items)) throw new Error("Keine Bitwarden-Einträge gefunden");
  return items.map((raw) => {
    const it = raw as Record<string, any>;
    return {
      title: it.name ?? "(ohne Titel)",
      username: it.login?.username ?? "",
      password: it.login?.password ?? "",
      url: it.login?.uris?.[0]?.uri ?? "",
      notes: it.notes ?? "",
      category: "Login",
    };
  });
}

function parseKeePassXml(text: string): NewEntry[] {
  if (text.charCodeAt(0) === 0x03) {
    throw new Error("Binäre .kdbx werden nicht unterstützt — als XML exportieren");
  }
  const doc = new DOMParser().parseFromString(text, "application/xml");
  if (doc.querySelector("parsererror")) throw new Error("Ungültige KeePass-XML");
  const out: NewEntry[] = [];
  doc.querySelectorAll("Entry").forEach((entry) => {
    const fields: Record<string, string> = {};
    entry.querySelectorAll(":scope > String").forEach((s) => {
      const key = s.querySelector("Key")?.textContent ?? "";
      const val = s.querySelector("Value")?.textContent ?? "";
      if (key) fields[key] = val;
    });
    if (!fields.Title && !fields.UserName && !fields.Password) return;
    out.push({
      title: fields.Title || "(ohne Titel)",
      username: fields.UserName || "",
      password: fields.Password || "",
      url: fields.URL || "",
      notes: fields.Notes || "",
      category: "Login",
    });
  });
  return out;
}

/** Minimal RFC-4180 CSV parser (handles quotes, escaped quotes, CRLF). */
function parseCsvRows(text: string): string[][] {
  const rows: string[][] = [];
  let row: string[] = [];
  let field = "";
  let inQuotes = false;
  for (let i = 0; i < text.length; i++) {
    const c = text[i];
    if (inQuotes) {
      if (c === '"') {
        if (text[i + 1] === '"') {
          field += '"';
          i++;
        } else inQuotes = false;
      } else field += c;
    } else if (c === '"') {
      inQuotes = true;
    } else if (c === ",") {
      row.push(field);
      field = "";
    } else if (c === "\n" || c === "\r") {
      if (c === "\r" && text[i + 1] === "\n") i++;
      row.push(field);
      field = "";
      if (row.length > 1 || row[0] !== "") rows.push(row);
      row = [];
    } else field += c;
  }
  if (field !== "" || row.length) {
    row.push(field);
    if (row.length > 1 || row[0] !== "") rows.push(row);
  }
  return rows;
}

function idx(headers: string[], aliases: string[]): number {
  return headers.findIndex((h) => aliases.includes(h.trim().toLowerCase()));
}

function parseCsv(text: string): NewEntry[] {
  const rows = parseCsvRows(text);
  if (!rows.length) throw new Error("Leere CSV");
  const headers = rows[0].map((h) => h.trim());
  const tI = idx(headers, ["name", "title", "account"]);
  const uI = idx(headers, ["username", "user", "login_username", "email", "login"]);
  const pI = idx(headers, ["password", "login_password", "pass"]);
  const lI = idx(headers, ["url", "uri", "website", "login_uri"]);
  const nI = idx(headers, ["notes", "note", "extra", "comments"]);
  if (pI < 0 && uI < 0) throw new Error("Keine Benutzername/Passwort-Spalten erkannt");

  const get = (r: string[], i: number) => (i >= 0 ? (r[i] ?? "").trim() : "");
  return rows.slice(1).map((r) => {
    const title = get(r, tI) || get(r, lI) || "(ohne Titel)";
    return {
      title,
      username: get(r, uI),
      password: get(r, pI),
      url: get(r, lI),
      notes: get(r, nI),
      category: "Login",
    };
  });
}

File: src-tauri/src/importer.rs
//! Importers for external password managers. All parsing happens locally on
//! bytes already read from disk — no network access.
//!
//! Supported:
//!   - Bitwarden (JSON export)
//!   - KeePass 2.x (unencrypted XML export)
//!   - CSV family via smart header mapping: Chrome/Edge, Firefox, LastPass,
//!     1Password (CSV), and generic CSV.
//!
//! KDBX (binary KeePass) is intentionally not parsed; users should export to
//! XML or CSV. We surface a clear message instead of failing opaquely.

use quick_xml::events::Event;
use quick_xml::Reader;
use serde::Deserialize;
use serde_json::Value;

use crate::vault::NewEntry;

#[derive(Debug, Clone, serde::Serialize)]
pub struct ImportResult {
    pub imported: usize,
    pub skipped: usize,
    pub entries: Vec<NewEntry>,
}

/// Parse bytes according to `format`. Returns the proposed entries (the caller
/// decides whether to commit them to the vault).
pub fn parse(format: &str, bytes: &[u8]) -> Result<Vec<NewEntry>, String> {
    match format.to_lowercase().as_str() {
        "bitwarden" => parse_bitwarden(bytes),
        "keepass" | "keepass-xml" => parse_keepass_xml(bytes),
        "lastpass" | "chrome" | "edge" | "firefox" | "1password" | "csv" | "generic" => {
            parse_csv(bytes)
        }
        other => Err(format!("unsupported import format: {other}")),
    }
}

// ---------------------------------------------------------------------------
// Bitwarden JSON
// ---------------------------------------------------------------------------

#[derive(Deserialize)]
struct BwExport {
    items: Vec<Value>,
}

fn parse_bitwarden(bytes: &[u8]) -> Result<Vec<NewEntry>, String> {
    let export: BwExport =
        serde_json::from_slice(bytes).map_err(|e| format!("invalid Bitwarden JSON: {e}"))?;
    let mut out = Vec::new();
    for item in export.items {
        // type 1 = login; we still import others as notes.
        let title = item
            .get("name")
            .and_then(Value::as_str)
            .unwrap_or("(ohne Titel)")
            .to_string();
        let notes = item
            .get("notes")
            .and_then(Value::as_str)
            .unwrap_or("")
            .to_string();
        let login = item.get("login");
        let username = login
            .and_then(|l| l.get("username"))
            .and_then(Value::as_str)
            .unwrap_or("")
            .to_string();
        let password = login
            .and_then(|l| l.get("password"))
            .and_then(Value::as_str)
            .unwrap_or("")
            .to_string();
        let url = login
            .and_then(|l| l.get("uris"))
            .and_then(Value::as_array)
            .and_then(|a| a.first())
            .and_then(|u| u.get("uri"))
            .and_then(Value::as_str)
            .unwrap_or("")
            .to_string();

        out.push(NewEntry {
            title,
            username,
            password,
            url,
            notes,
            category: "Login".into(),
        });
    }
    Ok(out)
}

// ---------------------------------------------------------------------------
// CSV family (smart header mapping)
// ---------------------------------------------------------------------------

fn first_match(headers: &[String], aliases: &[&str]) -> Option<usize> {
    for (i, h) in headers.iter().enumerate() {
        let hl = h.trim().to_lowercase();
        if aliases.iter().any(|a| hl == *a) {
            return Some(i);
        }
    }
    None
}

fn parse_csv(bytes: &[u8]) -> Result<Vec<NewEntry>, String> {
    let mut rdr = csv::ReaderBuilder::new()
        .flexible(true)
        .from_reader(bytes);
    let headers: Vec<String> = rdr
        .headers()
        .map_err(|e| format!("invalid CSV header: {e}"))?
        .iter()
        .map(|s| s.to_string())
        .collect();

    let title_idx = first_match(&headers, &["name", "title", "account"]);
    let user_idx = first_match(&headers, &["username", "user", "login_username", "email", "login"]);
    let pass_idx = first_match(&headers, &["password", "login_password", "pass"]);
    let url_idx = first_match(&headers, &["url", "uri", "website", "login_uri"]);
    let notes_idx = first_match(&headers, &["notes", "note", "extra", "comments"]);

    if pass_idx.is_none() && user_idx.is_none() {
        return Err("CSV has no recognizable username/password columns".into());
    }

    let get = |rec: &csv::StringRecord, idx: Option<usize>| -> String {
        idx.and_then(|i| rec.get(i)).unwrap_or("").trim().to_string()
    };

    let mut out = Vec::new();
    for result in rdr.records() {
        let rec = result.map_err(|e| format!("CSV row error: {e}"))?;
        let title = {
            let t = get(&rec, title_idx);
            if t.is_empty() {
                get(&rec, url_idx)
            } else {
                t
            }
        };
        out.push(NewEntry {
            title: if title.is_empty() {
                "(ohne Titel)".into()
            } else {
                title
            },
            username: get(&rec, user_idx),
            password: get(&rec, pass_idx),
            url: get(&rec, url_idx),
            notes: get(&rec, notes_idx),
            category: "Login".into(),
        });
    }
    Ok(out)
}

// ---------------------------------------------------------------------------
// KeePass 2.x XML
// ---------------------------------------------------------------------------

fn parse_keepass_xml(bytes: &[u8]) -> Result<Vec<NewEntry>, String> {
    // Reject binary KDBX early with a helpful message.
    if bytes.len() >= 4 && bytes[0] == 0x03 && bytes[1] == 0xD9 && bytes[2] == 0xA2 {
        return Err(
            "binary .kdbx files are not supported — export from KeePass as XML or CSV".into(),
        );
    }
    let text = std::str::from_utf8(bytes).map_err(|_| "KeePass XML is not valid UTF-8")?;
    let mut reader = Reader::from_str(text);
    reader.config_mut().trim_text(true);

    let mut out = Vec::new();
    let mut in_entry = false;
    let mut in_string = false;
    let mut cur_key: Option<String> = None;
    let mut last_tag = String::new();
    let mut pending_value = false;
    let mut entry = NewEntry {
        title: String::new(),
        username: String::new(),
        password: String::new(),
        url: String::new(),
        notes: String::new(),
        category: "Login".into(),
    };

    loop {
        match reader.read_event() {
            Ok(Event::Start(e)) => {
                let name = String::from_utf8_lossy(e.name().as_ref()).to_string();
                last_tag = name.clone();
                match name.as_str() {
                    "Entry" => {
                        in_entry = true;
                        entry = NewEntry {
                            title: String::new(),
                            username: String::new(),
                            password: String::new(),
                            url: String::new(),
                            notes: String::new(),
                            category: "Login".into(),
                        };
                    }
                    "String" if in_entry => {
                        in_string = true;
                        cur_key = None;
                    }
                    "Value" if in_string => pending_value = true,
                    _ => {}
                }
            }
            Ok(Event::Text(t)) => {
                let txt = t.unescape().unwrap_or_default().to_string();
                if in_string && last_tag == "Key" {
                    cur_key = Some(txt);
                } else if pending_value {
                    if let Some(key) = &cur_key {
                        match key.as_str() {
                            "Title" => entry.title = txt,
                            "UserName" => entry.username = txt,
                            "Password" => entry.password = txt,
                            "URL" => entry.url = txt,
                            "Notes" => entry.notes = txt,
                            _ => {}
                        }
                    }
                }
            }
            Ok(Event::End(e)) => {
                let name = String::from_utf8_lossy(e.name().as_ref()).to_string();
                match name.as_str() {
                    "Value" => pending_value = false,
                    "String" => in_string = false,
                    "Entry" => {
                        in_entry = false;
                        // Skip empty/recycle-bin artifacts.
                        if !(entry.title.is_empty()
                            && entry.username.is_empty()
                            && entry.password.is_empty())
                        {
                            if entry.title.is_empty() {
                                entry.title = "(ohne Titel)".into();
                            }
                            out.push(std::mem::replace(
                                &mut entry,
                                NewEntry {
                                    title: String::new(),
                                    username: String::new(),
                                    password: String::new(),
                                    url: String::new(),
                                    notes: String::new(),
                                    category: "Login".into(),
                                },
                            ));
                        }
                    }
                    _ => {}
                }
                last_tag.clear();
            }
            Ok(Event::Eof) => break,
            Err(e) => return Err(format!("KeePass XML parse error: {e}")),
            _ => {}
        }
    }
    Ok(out)
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn imports_bitwarden_json() {
        let json = br#"{
            "items": [
                {"name":"GitHub","notes":"n","login":{"username":"u","password":"p","uris":[{"uri":"https://github.com"}]}}
            ]
        }"#;
        let e = parse_bitwarden(json).unwrap();
        assert_eq!(e.len(), 1);
        assert_eq!(e[0].title, "GitHub");
        assert_eq!(e[0].username, "u");
        assert_eq!(e[0].password, "p");
        assert_eq!(e[0].url, "https://github.com");
    }

    #[test]
    fn imports_chrome_csv() {
        let csv = b"name,url,username,password\nGitHub,https://github.com,user,pass\n";
        let e = parse_csv(csv).unwrap();
        assert_eq!(e.len(), 1);
        assert_eq!(e[0].title, "GitHub");
        assert_eq!(e[0].username, "user");
        assert_eq!(e[0].password, "pass");
    }

    #[test]
    fn imports_lastpass_csv() {
        let csv = b"url,username,password,totp,extra,name,grouping,fav\nhttps://x.com,u@x.com,secret,,note,MyAccount,,0\n";
        let e = parse_csv(csv).unwrap();
        assert_eq!(e.len(), 1);
        assert_eq!(e[0].title, "MyAccount");
        assert_eq!(e[0].username, "u@x.com");
        assert_eq!(e[0].password, "secret");
        assert_eq!(e[0].notes, "note");
    }

    #[test]
    fn imports_keepass_xml() {
        let xml = br#"<?xml version="1.0"?>
        <KeePassFile><Root><Group><Entry>
            <String><Key>Title</Key><Value>Mail</Value></String>
            <String><Key>UserName</Key><Value>me@mail.com</Value></String>
            <String><Key>Password</Key><Value>pw123</Value></String>
            <String><Key>URL</Key><Value>https://mail.com</Value></String>
            <String><Key>Notes</Key><Value>hello</Value></String>
        </Entry></Group></Root></KeePassFile>"#;
        let e = parse_keepass_xml(xml).unwrap();
        assert_eq!(e.len(), 1);
        assert_eq!(e[0].title, "Mail");
        assert_eq!(e[0].username, "me@mail.com");
        assert_eq!(e[0].password, "pw123");
        assert_eq!(e[0].url, "https://mail.com");
        assert_eq!(e[0].notes, "hello");
    }

    #[test]
    fn rejects_binary_kdbx() {
        let kdbx = [0x03u8, 0xD9, 0xA2, 0x67, 0x00, 0x00];
        assert!(parse_keepass_xml(&kdbx).is_err());
    }
}
